src/Data/List/Properties.agda

```diff
@@ -753,10 +753,32 @@ module _ {P : A → Set p} (P? : Decidable P) where
   ... | no ¬px = contradiction eq (<⇒≢ (s≤s (length-filter xs)))
   ... | yes px = P.cong (x ∷_) (filter-complete (suc-injective eq))
 
+  filter-accept : ∀ {x xs} → P x → filter P? (x ∷ xs) ≡ x ∷ (filter P? xs)
+  filter-accept {x} Px with P? x
+  ... | yes _  = refl
+  ... | no ¬Px = contradiction Px ¬Px
+
+  filter-reject : ∀ {x xs} → ¬ P x → filter P? (x ∷ xs) ≡ filter P? xs
+  filter-reject {x} ¬Px with P? x
+  ... | yes Px = contradiction Px ¬Px
+  ... | no  _  = refl
+
+  filter-idem : filter P? ∘ filter P? ≗ filter P?
+  filter-idem []       = refl
+  filter-idem (x ∷ xs) with P? x | inspect P? x
+  ... | no  _ | _                   = filter-idem xs
+  ... | yes _ | P.[ eq ] rewrite eq = cong (x ∷_) (filter-idem xs)
+
```

Comment:
Alternatively this could be done with equational reasoning by using `filter-accept`.

Reply:
True and probably be more elegant, although it does require an extra pattern match and a bit more space. I think on balance I'm just going to leave it as is,